File: leetcode/q111_minimum_tree_depth.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def traverse(self, node, depth):
        if node == None:
            return
        
        if node.left == None and node.right == None:
            self.min_depth = min(self.min_depth, depth)
        
        self.traverse(node.left, depth + 1)
        self.traverse(node.right, depth + 1)
    
    def minDepth(self, root):
        self.min_depth = float('inf')
        """
        :type root: TreeNode
        :rtype: int
        """
        
        if root == None:
            return 0
        
        self.traverse(root, 0)
        return self.min_depth + 1
        

File: leetcode/Q111MinimumTreeDepth.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Q111MinimumTreeDepth {
    Double minDepth = Double.POSITIVE_INFINITY;
    
    public void traverse(TreeNode node, int depth) {
        if(node == null) {
            return;
        }
        
        if(node.left == null && node.right == null) {
            minDepth = Math.min(minDepth, (double) depth);
        }
        
        traverse(node.left, depth + 1);
        traverse(node.right, depth + 1);
    }
    
    public int minDepth(TreeNode root) {
        if(root == null) {
            return 0;
        }
        
        traverse(root, 1);
        return minDepth.intValue();
    }
}
